assets/js/components/higherorder/withData.js

```diff
@@ -153,100 +176,141 @@ const withData = (
 				moduleRequiringSetup: '',
 			};
 
+			this.handleModuleDataReset = this.handleModuleDataReset.bind( this );
+			this.handleReturnedData = this.handleReturnedData.bind( this );
+			this.addDataRequests = this.addDataRequests.bind( this );
+			this.removeDataRequests = this.removeDataRequests.bind( this );
+		}
+
+		componentDidMount() {
 			addAction(
 				'googlesitekit.moduleDataReset',
 				'googlesitekit.moduleDataResetHandler',
-				() => {
-					this.setState( {
-						data: false,
-						errorMessage: false,
-						zeroData: false,
-					} );
+				this.handleModuleDataReset
+			);
+
+			this.addDataRequests();
+		}
+
+		componentWillUnmount() {
+			removeAction(
+				'googlesitekit.moduleDataReset',
+				'googlesitekit.moduleDataResetHandler',
+				this.handleModuleDataReset
+			);
+
+			this.removeDataRequests();
+		}
+
+		handleModuleDataReset() {
+			// When the global dateRange changes, it will trigger the googlesitekit.moduleDataReset action.
+			// When this happens, we need to re-hook the requests for the default date range to be applied correctly.
+			this.removeDataRequests();
+			this.addDataRequests();
+
+			this.setState( {
+				data: false,
+				errorMessage: false,
+				zeroData: false,
+			} );
+		}
+
+		addDataRequests() {
+			const { dateRange } = this.props;
+
+			Object.entries( dataRequestsByContext ).forEach(
+				( [ context, dataRequests ] ) => {
+					addFilter(
+						`googlesitekit.module${ context }DataRequest`,
+						`googlesitekit.withData.${ hashRequests( dataRequests ) }`,
+						( contextRequests ) => {
```

Comment:
This is also a bit more efficient than before. Previously we were adding a filter callback for each request, for each of its contexts. Now we only add a filter once for each _context_ and add all relevant requests at once, for each wrapped component.